prototype/galore/galore_fused/triton/kernels/adam_step.py

```diff
@@ -0,0 +1,178 @@
+import torch
+import triton
+import triton.language as tl
+from triton.language.math import sqrt
+from triton.runtime.autotuner import heuristics
+
+from galore_fused.triton.custom_autotune import Config, autotune
+
+BETA1, BETA2 = 0.9, 0.999
+EPS = 1e-8
+
+AUTOTUNER_TOP_K = 10
+
+
+def get_configs_for_adam(num_warps=[2, 4, 8], block_sizes=[512, 1024, 2048]):
+    configs = []
+    for w in num_warps:
+        for bs in block_sizes:
+            configs.append(Config({"BLOCK_SIZE": bs}, num_warps=w))
+    return configs
+
+
+def early_adam_prune(configs, named_args):
+    numels = named_args["numels"]
+    pruned_configs = [cfg for cfg in configs if numels % cfg.kwargs["BLOCK_SIZE"] == 0]
+    # print("Pruned configs:\n")
+    for cfg in pruned_configs:
+        print(f"{cfg}\n")
+    return pruned_configs
+
+
+def get_adam_tuner(
+    configs=get_configs_for_adam(),
+    early_config_prune=None,  # early_adam_prune,
+    top_k=AUTOTUNER_TOP_K,
+):
+    return autotune(
+        configs=configs,
+        prune_configs_by={
+            "early_config_prune": early_config_prune,
+            "top_k": top_k,
+        },
+        key=["numels"],
+    )
+
+
+def get_adam_heuristics():
+    return {
+        "USE_MASK": lambda args: args["numels"] % args["BLOCK_SIZE"] != 0,
+    }
+
+
+@autotune(configs=get_configs_for_adam(), key=["numels"])
```

Comment:
very clean kernel - @janeyx99 I'm guessing this is also something `torch.compile` can codegen just being less human readable

I guess overall I'm having trouble figuring out whether we should be merging kernels in ao if torch.compile can codegenerate them